code_generation/quantity.py

```diff
@@ -26,39 +38,106 @@ def reserve_scope(self, scope):
+        Returns:
+            list. List of leafs
+        """
+        result = [self.name] + [
+            self.get_leaf(shift, scope) for shift in self.get_shifts(scope)
+        ]
+        return result
 
     def shift(self, name, scope):
+        """
+        Function to define a shift for a given scope. If the shift is marked as ignored, nothing will be added.
+        When a new shift is defined, all child quantities of a given quantity will be shifted as well.
+        Shifts defined in the global scope, are added to the shift dictionary of all scopes.
+        Shifts that are exclusive to a single scope are only added to the given scope.
+
+        Args:
+            name (str): Name of the shift
+            scope (str): Scope for which the shift should be defined
+        Returns:
+            None
+        """
         if scope in self.ignored_shifts.keys():
             if name in self.ignored_shifts[scope]:
                 log.debug("Ignoring shift {} for quantity {}".format(name, self.name))
                 return
         log.debug("Adding shift {} to quantity {}".format(name, self.name))
-        if not scope in self.shifts.keys():
-            self.shifts[scope] = set()
-        if not name in self.shifts[scope]:
-            self.shifts[scope].add(name)
-            if scope == "global":  # shift children in all scopes if scope is global
-                for any_scope in self.children:
-                    for c in self.children[any_scope]:
-                        c.shift(name, any_scope)
+        # adding new shifts to scopes:
+        # if a shift is defined for the global scope, it should also be added for all other scopes,
+        # therefore, if a new scope is added,
+        # this scope is a copy of the global scope.
+        if scope not in self.shifts.keys():
+            if "global" in self.shifts.keys():
+                # make a copy of the global shifts
+                self.shifts[scope] = self.shifts["global"]
             else:
-                if scope in self.children.keys():
-                    for c in self.children[scope]:
-                        c.shift(name, scope)
+                self.shifts[scope] = {}
+        scopes = [scope]
+        if scope == "global" and scope in self.shifts.keys():
+            # in this case, we add the shift to all existing scopes
+            scopes = self.shifts.keys()
+        for scope in scopes:
+            if name not in self.shifts[scope]:
+                self.shifts[scope][name] = self.name + name
+                if scope == "global":  # shift children in all scopes if scope is global
+                    for any_scope in self.children:
+                        for c in self.children[any_scope]:
+                            c.shift(name, any_scope)
+                else:
+                    if scope in self.children.keys():
+                        for c in self.children[scope]:
+                            c.shift(name, scope)
```

Comment:
This should not be done within the loop for each scope but only for the initial one.
After moving it outside the loop, mind the conflicting variable name `scope` inside and outside the loop.

Reply:
quantity class changes are reversed